pkg/pyramid/tier_fs.go

```
@@ -156,6 +156,14 @@ func (tfs *TierFS) store(ctx context.Context, storageID, namespace, originalPath
 	if err != nil {
 		return fmt.Errorf("open file %s: %w", originalPath, err)
 	}
+	defer func() {
+		if err != nil {
+			// If there is an error, remove the temp file (in the happy path, it's removed in the end of the "store" func)
+			if removeErr := os.Remove(originalPath); removeErr != nil {
+				err = fmt.Errorf("%w and remove tempfile: %w", err, removeErr)
```

Without named return (`(err error)` in the return position) this won't work, because the deferred func captures the error as it is at l.159.  See [playgroud](https://go.dev/play/p/nqG2jubeFAm)

---

Not sure I follow -
Plus, it seems to work in the unit tests...

Can you give a concrete example in this context?

---

So I double-checked, and I got it wrong! But not in the way you probably think 🕵 
When the error is a local variable, the `if err != nil` condition _does_ work, but reassigning the error doesn't.
This is because defer statements are only allowed to modify named return values.  Because Go is _that_ simple and intuitive - no way to get it wrong!  ([source](https://boldlygo.tech/archive/2024-08-27-modifying-return-values-in-deferred-functions/) + [modified playground](https://go.dev/play/p/mXsc3bg1Hhy))

In your case, it means the `os.Remove` call indeed runs, but if it fails, you won't see it in the error chain.

So, still blocking because l.163 is meaningless.  Either use named return values or simply ignore the os.Remove error.
IMO ignoring the error is OK in this case, as it's has minimal effect either way.